segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,523 @@
+        /**
+         * Returns a {@link CompletableFuture} that results from the execution of the following tasks for each
+         * {@link ContainerEventProcessor.EventProcessor}: i) Read available data in the Segment, ii) deserialize at most
+         * {@link ContainerEventProcessor.EventProcessorConfig#getMaxItemsAtOnce()} events, iii) execute the handler
+         * function for the {@link ContainerEventProcessor.EventProcessor} on the event list, and iv) truncate the
+         * internal Segment.
+         */
+        private CompletableFuture<Void> processEvents() {
+            final Timer iterationTime = new Timer();
+            return readEvents()
+                    .thenComposeAsync(this::applyProcessorHandler, this.executor)
+                    .thenComposeAsync(readResult -> truncateInternalSegment(readResult, iterationTime), this.executor)
+                    .handleAsync((r, ex) -> {
+                        // If we got an exception different from NoDataAvailableException, report it as something is off.
+                        if (ex != null && !(Exceptions.unwrap(ex) instanceof NoDataAvailableException)) {
+                            log.warn("{}: Processing iteration failed, retrying.", this.traceObjectId, ex);
+                            this.failedIteration.set(true);
+                        } else {
+                            this.failedIteration.set(false);
+                        }
+                        return null;
+                    });
+        }
+
+        /**
+         * Attempts to read data from the internal Segment. The size of the read length would be the minimum between the
+         * outstanding bytes and 2 * ProcessorEventData.MAX_TOTAL_EVENT_SIZE bytes.
+         *
+         * @return A {@link CompletableFuture} that, when completed, will contain a {@link EventsReadAndTruncationLength}
+         * object with events read and the next truncation length for the internal Segment.
+         */
+        private CompletableFuture<EventsReadAndTruncationLength> readEvents() {
+            return CompletableFuture.supplyAsync(() -> {
+                                        int readLength = (int) Math.min(getOutstandingBytes(), 2 * ProcessorEventData.MAX_EVENT_SIZE);
+                                        return this.segment.read(this.segment.getInfo().getStartOffset(), readLength, containerOperationTimeout);
+                                    }, this.executor)
+                                    .thenCompose(rr -> AsyncReadResultProcessor.processAll(rr, this.executor, containerOperationTimeout))
+                                    .thenApply(this::deserializeEvents);
+        }
+
+        @SneakyThrows(Exception.class)
+        private EventsReadAndTruncationLength deserializeEvents(BufferView inputData) {
+            List<ProcessorEventData> events = new ArrayList<>();
+            long truncationLength = 0;
+            Exception deserializationException = new NoDataAvailableException();
```

Comment:
Creating an exception every time you call this is expensive. It will need to unwind the stack trace and record it. It will slow your program down.

Please only create exceptions when you are ready to throw them or complete a Future with it.

Reply:
Done.